## Librerías

In [1]:
# Para visualizar más de una salida de una celda
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

## Preguntas

### Pregunta 1

Escriba un código en Python, usando comprensión de listas, que le permita encontrar los números entre 1 y 1000 (inclusivos), que tengan al menos un dígito '3' en ellos.

In [6]:
arr = []
for num in range(1,1001):
    if ('3' in str(num)):
        arr.append(num)
print(arr)


[3, 13, 23, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 43, 53, 63, 73, 83, 93, 103, 113, 123, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 143, 153, 163, 173, 183, 193, 203, 213, 223, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 243, 253, 263, 273, 283, 293, 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321, 322, 323, 324, 325, 326, 327, 328, 329, 330, 331, 332, 333, 334, 335, 336, 337, 338, 339, 340, 341, 342, 343, 344, 345, 346, 347, 348, 349, 350, 351, 352, 353, 354, 355, 356, 357, 358, 359, 360, 361, 362, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 373, 374, 375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387, 388, 389, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 403, 413, 423, 430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 443, 453, 463, 473, 483, 493, 503, 513, 523, 530, 531, 532, 533, 534, 535, 536, 537, 538, 539, 543, 553, 563, 573, 583, 593, 603, 613, 623, 630, 631, 632, 633, 634, 635,

### Pregunta 2

Escriba un código en Python, usando listas por comprensión, que encuentre los números en común de dos listas (sin usar tuplas o sets)

Use las siguientes listas para probar su código:

list_a = [1, 2, 3, 4]

list_b = [2, 3, 4, 5]

In [8]:
list_a = [1, 2, 3, 4]
list_b = [2, 3, 4, 5]

arr = []
for elem in list_a:
    if elem in list_b:
        arr.append(elem)
arr

[2, 3, 4]

### Pregunta 3

Complete el siguiente código Python para: 

Crear una función func_compute(n) que retorne lo producido por una función lambda. La función func_compute(n) toma n y lo multiplica un número desconocido x.
Complete también el resto del código. 

El resultado debe verse así:

Dobla el número 15 = 30
Triplica el número 15 = 45
Cuadriplica el número 15 = 60

In [11]:
def func_compute(n):
    return lambda x: n*x

resultado = func_compute(2)
print("Duplica el número 15 =", resultado(15)) 

resultado = func_compute(3) 
print("Triplica el número 15 =", resultado(15)) 

resultado = func_compute(4)
print("Cuadriplica el número 15 =", resultado(15)) 

Duplica el número 15 = 30
Triplica el número 15 = 45
Cuadriplica el número 15 = 60


### Pregunta 4

Complete el siguiente código Python que calcula el cuadrado y el cubo todos los números de la siguiente lista de enteros usando funciones lambda.

El resultado (salida del código) debe verse así:

Lista original: [1,2,3,4,5,6,7,8,9,10]

Lista al cuadrado: [1,4,9, 16, 25, 36, 49, 64, 81, 100]

Lista al cubo: [1, 8, 27, 64, 125, 216, 343, 512, 729, 1000]

In [14]:
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] 

print("Lista original:") 
print(nums) 
print("\nLista al cuadrado:") 
square_nums = list(map(lambda x: x ** 2, nums))


print(square_nums) 
print("\nLista al cubo:") 
cube_nums = list(map(lambda x: x ** 3, nums))

 
print(cube_nums)

Lista original:
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

Lista al cuadrado:
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

Lista al cubo:
[1, 8, 27, 64, 125, 216, 343, 512, 729, 1000]


### Pregunta 5

Escribe un programa Python para encontrar números de una lista que sean divisibles entre 19 o entre 13, usando una función lambda.

El resultado debería verse así:

Lista original: [19, 65, 57, 39, 152, 639, 121, 44, 90, 190]

Números que son divisibles entre 19 o 13: [19, 65, 57, 39, 152, 190]

In [17]:
lista = [19, 65, 57, 39, 152, 639, 121, 44, 90, 190]
resultado = list(filter(lambda x: x%19==0 or x%13==0, lista))

print('Lista original: ', lista)
print('Números que son divisibles entre 19 o 13: ', resultado)

Lista original:  [19, 65, 57, 39, 152, 639, 121, 44, 90, 190]
Números que son divisibles entre 19 o 13:  [19, 65, 57, 39, 152, 190]


### Pregunta 6

Complete la función imp_lineas(arch,subcad) a la que se le pasa un nombre de archivo (incluida su ruta) y una cadena e imprima solo las líneas del archivo que contengan la cadena. Dicha función debe llamar a la siguiente función generadora:


In [27]:
def readfiles(filenames):   
    ''' filenames: lista de nombres de archivos con su ruta'''
    for f in filenames:
        for line in open(f):
            yield line
 

def imp_lineas(arch,subcad):
    '''
    Parameters
    ----------
    arch : list
        lista de nombres de archivos
    subcad : str
        cadena de caracteres presente en la línea a imprimir.
    '''

    files = readfiles(arch)
    r_lines = (line.strip() for line in files if subcad in line)

    [print(line) for line in r_lines]
    
       

archivos = ['texto1.txt', 'texto2.txt'] 

imp_lineas(archivos,'que')

hola que tal
creo que voy a reprobar
porque estoy muy idiota
que sorpresa
que decepcionante soy
no sé por que existo


### Pregunta 7

Completa la  función tot_lineas_ejecut(list_ipynb) que retorna el número total de líneas de cógido, ignorando líneas vacías y de comentario, en todos los archivos .ipynb de la lista que se le pase como argumento. Esta función debe llamar a la función generadora siguiente:

In [88]:
def readfiles(filenames):
    ''' filenames: lista de nombres de archivos con su ruta''' 
    for f in filenames: 
        for line in open(f): 
            yield line

def tot_lineas_ejecut(list_ipynb):
    '''
    list_ipynb:list
    Retorna el número total de líneas de cógido, ignorando líneas vacías y de comentario
    '''
    tot = 0

    files = readfiles(list_ipynb)

    for line in files:
        if line.strip() and not line.strip().startswith('#'):
            tot +=1
    return tot


### Pregunta 8

Complete la siguiente función generadora que provee valores que aumenten de 3 en 3, desde un valor  mínimo inclusivo(start) hasta un valor máximo exclusivo (stop). Escriba a continuación un código con argumentos de prueba que llame a genera_serie3() e imprime cada valor generado.

def genera_serie3(start, stop):
   ...

In [97]:
def genera_serie3(start, stop):
    cont = start
    for i in range(stop):
        if (i<=stop):
            yield cont
            cont += 3
        else:
            return

resultado = list(genera_serie3(0,10))
type(resultado)
[print(num) for num in resultado]

list

0
3
6
9
12
15
18
21
24
27


[None, None, None, None, None, None, None, None, None, None]

### Pregunta 9

Completa el siguiente código:



La salida debe verse así, incluyendo el texo:

Arreglo original: [1, 2, 3, 4]

Arreglo convertido a flotante: [ 1. 2. 3. 4.]

In [90]:
import numpy as np 
a = [1, 2, 3, 4] 
print("Arreglo original: ",a) 
x = np.array(a,dtype=float)
print("Arreglo convertido a flotante: ",x) 

Arreglo original:  [1, 2, 3, 4]
Arreglo convertido a flotante:  [1. 2. 3. 4.]


Hay muchas formas de hacerlo. Veamos algunas:
+ x = np.asfarray(a) 
+ x = np.array([float(i) for i in a]) 
+ x = np.asarray(a).astype('float') 
+ x = np.asarray(a,dtype='float') 

### Pregunta 10

Complete el código para crear un arreglo con valores de 12 a 38

import numpy as np 
_________________
print(x)
La salida al imprimir el arreglo debe verse así:

[12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37]

### Pregunta 11

Complete el código para indicar la posición del elemento de máximo valor  y del elemento de mínimo valor de un arreglo de numpy.

Si el arreglo original es [4,66,0,10,1], la salida debe verse así:

índice del máximo: 1
índice del mínimo: 2

Si el arreglo original es [[4,66,0],[10,68,3]], la salida debe verse así:

índice del máximo: 4
índice del mínimo: 2



In [92]:
# No se necesita usar flatten
import numpy as np 
x = np.array([[4,66,0],[10,68,3]]) 
x.flatten()
print("Arreglo original: ",x) 
print("índice del máximo: ", np.argmax(x)) 
print("índice del mínimo: ",np.argmin(x))

array([ 4, 66,  0, 10, 68,  3])

Arreglo original:  [[ 4 66  0]
 [10 68  3]]
índice del máximo:  4
índice del mínimo:  2


### Pregunta 12

Selecciona todas las opciones correctas para eliminar “Naranja” del set

sampleSet = {'Amarillo', 'Naranja', 'Negro'}

Grupo de opciones de respuesta

+ del sampleSet [“Naranja”]       (no se puede eliminar de un set)

+ sampleSet.remove('Naranja')     (sí)

+ sampleSet.pop(“Naranja”)        (Pop no toma argumentos)

+ sampleSet.discard(“Naranja”)    (sí) 

In [32]:
sampleSet = {'Amarillo', 'Naranja', 'Negro'}

# del sampleSet ['Naranja']
# sampleSet

sampleSet.discard('Naranja')
sampleSet


{'Amarillo', 'Negro'}

### Pregunta 13

Selecciona la manera correcta de copiar 2 sets

+ (X) set2 = set1.copy() 
+ ( ) set2 = set1(set1) 
+ ( ) set2.update(set1)
+ ( ) set1 = set1

In [39]:
set1 = {'Amarillo', 'Naranja', 'Negro'}
set2 = {}

# (X) set2 = set1.copy() 
# set2 = set1(set1) # Set obj not callable
# set2.update(set1) # Ni idea
# set2 = set1

id(set1)
set1
id(set2)
set2


2905026611232

{'Amarillo', 'Naranja', 'Negro'}

2905026611232

{'Amarillo', 'Naranja', 'Negro'}

### Pregunta 14

¿Cuál es el resultado del siguiente código?  
R: Amarillo

In [40]:
aTuple = 'Amarillo', 20, 'Rojo'
a, b, c = aTuple
print(a)

Amarillo


### Pregunta 15

Selecciona lo que es verdad sobre las tuplas en Python

+ (X) No podemos cambiar la tupla una vez creada

+ ( ) Podemos cambiar la tupla una vez creada

+ (X) La tupla no tiene orden

+ ( ) La tupla mantiene el orden de los ítems

### Pregunta 16

¿Cuál es el resultado del siguiente código?

In [43]:
dict1 = {'key':1,'key2':2}
dict2 = {'key':2,'key2':1}

print(dict1==dict2)


False


### Pregunta 17

Escribe en Python la manera correcta para acceder el valor de "historia"

![Alt text](image.png)

In [87]:
sampleDict = {
    'clase':{
        'estudiante':{
            'nombre':'Mike',
            'marks':{
                'historia':80
            }
        },
        'estudiante1':{
            'nombre':'laura',
            'marks':{
                'historia':80
            }
        }
    }
}

for key,estudiante in sampleDict['clase'].items():
    if estudiante['nombre'] == 'Mike':
        print(estudiante['marks']['historia'])
        break

80


### Pregunta 18

Escriba código Python usando la biblioteca pandas y otras requeridas para:

1- Crear un DataFrame con las columnas llamadas "x", "y" y "prom" 

2- Las columnas "x" , "y" contendrán valores aleatorios entre (5,10) y representan calificaciones de alumnos

3- La columna tendrá el promedio de las columnas x" , "y"

4- Guardará en un archivo binario los valores de la columna "prom" y en un archivo csv el DataFrame, sin que se guarde una columna con el índice.

In [75]:
import pandas as pd
import numpy as np
alumnos = 5

df = pd.DataFrame(data={'x':[],'y':[],'prom':[]})
df
df['x'] = np.random.randint(5,10, alumnos)
df['y'] = np.random.randint(5,10, alumnos)
df
df['prom'] = df.mean(axis=1)
df
 

tipos = pd.DataFrame(df.dtypes,columns=['dtypes'])
tipos.to_csv('tipos_pobreza.csv',index_label='columnas',header=['tipos'],encoding='utf_8')

proms = df['prom'].to_numpy()
np.save('promedio.npy', proms)
df.to_csv('calificaciones.csv',index=False,encoding='utf-8')



Empty DataFrame
Columns: [x, y, prom]
Index: []

x  y  prom
0  7  6   NaN
1  6  8   NaN
2  7  9   NaN
3  5  6   NaN
4  5  5   NaN

x  y  prom
0  7  6   6.5
1  6  8   7.0
2  7  9   8.0
3  5  6   5.5
4  5  5   5.0